Userland/Libraries/LibWeb/HTML/HTMLFormElement.cpp

```diff
@@ -319,6 +319,32 @@ WebIDL::ExceptionOr<void> HTMLFormElement::submit()
     return submit_form(*this, { .from_submit_binding = true });
 }
 
+// https://html.spec.whatwg.org/multipage/forms.html#dom-form-requestsubmit
+WebIDL::ExceptionOr<void> HTMLFormElement::request_submit(JS::GCPtr<Element> submitter)
+{
+    auto submitter_html_element = TRY([&]() -> WebIDL::ExceptionOr<JS::NonnullGCPtr<HTMLElement>> {
+        // 1. If submitter is not null, then:
+        if (submitter) {
+            // 1. If submitter is not a submit button, then throw a TypeError.
+            auto* form_associated_element = dynamic_cast<FormAssociatedElement*>(submitter.ptr());
+            if (!(form_associated_element && form_associated_element->is_submit_button()))
+                return vm().throw_completion<JS::TypeError>(JS::ErrorType::NotAnObjectOfType, "Submit button");
+
+            // 2. If submitter's form owner is not this form element, then throw a "NotFoundError" DOMException.
+            if (form_associated_element->form() != this)
+                return WebIDL::NotFoundError::create(realm(), "The submitter is not owned by this form element"_fly_string);
+
+            return JS::NonnullGCPtr { form_associated_element->form_associated_element_to_html_element() };
+        }
+
+        // 2. Otherwise, set submitter to this form element.
+        return JS::NonnullGCPtr { *this };
+    }());
```

Comment:
Not sure I see why we need an IIFE. Does the following work?
```c++
// https://html.spec.whatwg.org/multipage/forms.html#dom-form-requestsubmit
WebIDL::ExceptionOr<void> HTMLFormElement::request_submit(JS::GCPtr<Element> submitter)
{
    // 1. If submitter is not null, then:
    if (submitter) {
        // 1. If submitter is not a submit button, then throw a TypeError.
        auto* form_associated_element = dynamic_cast<FormAssociatedElement*>(submitter.ptr());
        if (!(form_associated_element && form_associated_element->is_submit_button()))
            return WebIDL::SimpleException { WebIDL::SimpleExceptionType::TypeError, "The submitter is not a submit button"sv };

        // 2. If submitter's form owner is not this form element, then throw a "NotFoundError" DOMException.
        if (form_associated_element->form() != this)
            return WebIDL::NotFoundError::create(realm(), "The submitter is not owned by this form element"_fly_string);
    }
    // 2. Otherwise, set submitter to this form element.
    else {
        submitter = this;
    }

    // 3. Submit this form element, from submitter.
    return submit_form(static_cast<HTMLElement&>(*submitter), {});
}
```

Reply:
That's much nicer, thanks!